src/com/zone5cloud/http/core/Z5HttpClient.java

```diff
@@ -214,7 +226,7 @@ private void refreshTokenIfRequired() throws InterruptedException {
 					String username = token.extractUsername();
 					if (username != null) {
 						OAuthTokenRequest request = new OAuthTokenRequest();
-						request.setUsername(token.extractUsername());
+						request.setUsername(this.userName);
 						request.setRefreshToken(token.getRefreshToken());
 						request.setClientId(clientID);
 						request.setClientSecret(clientSecret);
```

Comment:
You are still pulling username out of token.extractUsername() and checking that for null. In the case where username is not in the token this code will never be executed. You need line 226 to be:
`String username = this.userName;`
and then line 229 to be:
`request.setUsername(username);`